icu4c/source/common/uniquecharstr.h

```diff
@@ -48,21 +50,41 @@ class UniqueCharStrings {
 
     /**
      * Adds a string and returns a unique number for it.
-     * The string's buffer contents must not change, nor move around in memory,
-     * while this UniqueCharStrings is in use.
-     * The string contents must be NUL-terminated exactly at s.length().
-     *
+     * If copy is false, then the string's buffer contents must not change,
+     * nor move around in memory, while this UniqueCharStrings is in use.
      * Best used with read-only-alias UnicodeString objects that point to
      * stable storage, such as strings returned by resource bundle functions.
+     * The string contents must be NUL-terminated exactly at s.length().
+     *
+     * If copy is true, a copy of the string will be stored and freed internally.
      */
-    int32_t add(const UnicodeString &s, UErrorCode &errorCode) {
+    int32_t add(const UnicodeString &s, bool copy, UErrorCode &errorCode) {
         if (U_FAILURE(errorCode)) { return 0; }
         if (isFrozen) {
             errorCode = U_NO_WRITE_PERMISSION;
             return 0;
         }
         // The string points into the resource bundle.
-        const char16_t *p = s.getBuffer();
+        const char16_t *p = nullptr;
+        // If s cannot be guaranteed to stay allocated, we copy and track the
+        // value internally.
+        if (copy) {
+            UnicodeString* cloned = s.clone();
+            if (cloned == nullptr) {
+                errorCode = U_MEMORY_ALLOCATION_ERROR;
+                return 0;
+            }
+            // Guarantee the p is pointed to a NUL-terminated string.
+            p = cloned->getTerminatedBuffer();
+            copied.adoptElement(cloned, errorCode);
+        } else {
+            p = s.getBuffer();
+            // The string contents must be NUL-terminated exactly at s.length().
+            if (p[s.length()] != '\0') {
+                errorCode = U_ILLEGAL_ARGUMENT_ERROR;
+                return 0;
+            }
+        }
```

Comment:
Actually, why do you need to do any of this? `s` is only ever used as an argument to `strings->appendInvariantChars`. It's not clear why the string needs to be NUL terminated when passing to that function.

Maybe the call to `strings->append(0, errorCode);` a few lines down needs to be moved after the call to `strings->appendInvariantChars`.

Reply:
The issue is uhash_puti(&map, const_cast<char16_t *>(p)... )
the map is initialized with  uhash_init(&map, uhash_hashUChars, uhash_compareUChars, uhash_compareLong, &errorCode);
so p need to be processed by uhash_hashUChars and uhash_compareUChars which both expect "null-terminated UChar* string"

```
/**
 * Generate a hash code for a null-terminated UChar* string.  If the
 * string is not null-terminated do not use this function.  Use
 * together with uhash_compareUChars.
 * @param key The string (const UChar*) to hash.
 * @return A hash code for the key.
 */
U_CAPI int32_t U_EXPORT2
uhash_hashUChars(const UHashTok key);
...
/**
 * Comparator for null-terminated UChar* strings.  Use together with
 * uhash_hashUChars.
 * @param key1 The string for comparison
 * @param key2 The string for comparison
 * @return true if key1 and key2 are equal, return false otherwise.
 */
U_CAPI UBool U_EXPORT2
uhash_compareUChars(const UHashTok key1, const UHashTok key2);
```